File: backend/schemas/smart_input.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from core.enums import SmartInputMode, SmartInputStatus, TransactionType
from schemas.category import CategoryResponse
from schemas.transaction import TransactionResponse


class VoiceDraftCreate(BaseModel):
    raw_text: str = Field(min_length=1)


class SmartInputDraftUpdate(BaseModel):
    parsed_type: TransactionType | None = None
    parsed_amount_minor: int | None = Field(default=None, gt=0)
    parsed_description: str | None = Field(default=None, min_length=1, max_length=255)
    suggested_category_id: UUID | None = None
    merchant_name: str | None = Field(default=None, min_length=1, max_length=150)


class SmartInputConfirmRequest(BaseModel):
    transaction_date: date | None = None


class SmartInputDraftResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    mode: SmartInputMode
    status: SmartInputStatus
    raw_text: str | None
    source_file_ref: str | None
    parsed_type: TransactionType | None
    parsed_amount_minor: int | None
    parsed_description: str | None
    merchant_name: str | None
    confidence_percent: int | None
    parser_payload: dict | None
    expires_at: datetime | None
    created_at: datetime
    updated_at: datetime
    suggested_category: CategoryResponse | None


class SmartInputConfirmResponse(BaseModel):
    draft_id: UUID
    status: SmartInputStatus
    transaction: TransactionResponse

File: frontend/src/app/lib/api/smartInput.ts
import { apiRequest } from "./client";
import { TransactionResponse } from "./transactions";
import { CategoryResponse } from "./categories";

export type SmartInputMode = "voice" | "ocr";
export type SmartInputStatus = "processing" | "draft" | "confirmed" | "failed" | "discarded";

export interface SmartInputDraftResponse {
  id: string;
  mode: SmartInputMode;
  status: SmartInputStatus;
  raw_text: string | null;
  source_file_ref: string | null;
  parsed_type: "income" | "expense" | null;
  parsed_amount_minor: number | null;
  parsed_description: string | null;
  merchant_name: string | null;
  confidence_percent: number | null;
  parser_payload: any | null;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
  suggested_category: CategoryResponse | null;
}

export interface SmartInputDraftUpdate {
  parsed_type?: "income" | "expense";
  parsed_amount_minor?: number;
  parsed_description?: string;
  suggested_category_id?: string;
  merchant_name?: string;
}

export interface SmartInputConfirmRequest {
  transaction_date?: string; // ISO date string YYYY-MM-DD
}

export interface SmartInputConfirmResponse {
  draft_id: string;
  status: SmartInputStatus;
  transaction: TransactionResponse;
}

/**
 * Create a smart-input draft from an audio file (ASR)
 */
export async function createVoiceDraft(file: File): Promise<SmartInputDraftResponse> {
  const formData = new FormData();
  formData.append("file", file);

  return apiRequest<SmartInputDraftResponse>("/smart-input/drafts/voice", {
    method: "POST",
    body: formData,
  });
}

/**
 * Create a smart-input draft from an image file (OCR)
 */
export async function createOcrDraft(file: File, hintText?: string): Promise<SmartInputDraftResponse> {
  const formData = new FormData();
  formData.append("file", file);
  if (hintText) {
    formData.append("hint_text", hintText);
  }

  return apiRequest<SmartInputDraftResponse>("/smart-input/drafts/ocr", {
    method: "POST",
    body: formData,
  });
}

/**
 * Get a specific smart-input draft by ID
 */
export async function getSmartInputDraft(draftId: string): Promise<SmartInputDraftResponse> {
  return apiRequest<SmartInputDraftResponse>(`/smart-input/drafts/${draftId}`, {
    method: "GET",
  });
}

/**
 * Update a smart-input draft (e.g., user corrects the amount or category)
 */
export async function updateSmartInputDraft(
  draftId: string,
  payload: SmartInputDraftUpdate
): Promise<SmartInputDraftResponse> {
  return apiRequest<SmartInputDraftResponse>(`/smart-input/drafts/${draftId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

/**
 * Confirm a draft and turn it into a real transaction
 */
export async function confirmSmartInputDraft(
  draftId: string,
  payload: SmartInputConfirmRequest = {}
): Promise<SmartInputConfirmResponse> {
  return apiRequest<SmartInputConfirmResponse>(`/smart-input/drafts/${draftId}/confirm`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
